### ISSUE TITLE
Ramda splat packages

### ISSUE THREAD

**Issue by krainboltgreene** [ramda/ramda#2151]
One thing I'm currently getting ready to do with https://github.com/krainboltgreene/ramda-extra.js/tree/core/source is to create a package for each individual function in addition to the main collectors package. It recently got pretty easy with lerna. Any interest in ramda crew having the same?

**Comment by CrossEye** [comment]
Sorry for the slow reply; I've been mostly offline for some weeks.

I know that's the way lodash has gone too.  I haven't really figured out the advantages.   Can you explain why you're doing that?

**Comment by krainboltgreene** [comment]
So I've now done it for ramda-extra, you can see the organization here: https://www.npmjs.com/org/unction

Here's what I've learned:

  - The tooling is like 80% there. Lerna gets close, but it has some bad choices that make it a pain.
  - Organizations are nice, but require you to publish with `--access=public` which can be annoying. There are some weird error messages associated with this.
  - A single repo for multiple packages is OK, but it's kinda hard to do both a collection package (`@unction/all` basically doesn't work) and also do this splat package thing.

Here's the value add: **Change is significantly easier**. If I need ramda for `path()` and there's a newer version that's more performant I now have to pay attention to breaking changes. Ramda isn't currently following semantic versioning (which I find to be very awkward anyways) but at least with splat packages you know exactly what you're getting. Deprecating a function? No problem! People can still download that package but get performance improvements without having to rewrite whole swaths of code.

**Comment by CrossEye** [comment]
Thank you.  I need to think about this a bit.  I'm hoping others with any experience might chime in here too.

I haven't had to do the smaller imports very often, but I do like the idea.  What I don't like is importing multiple copies of shared dependencies.  But I suppose users can balance that for themselves.  Three or four Ramda imports might make sense, but if they need a dozen, then they should probably just import the whole thing.

**Comment by krainboltgreene** [comment]
> What I don't like is importing multiple copies of shared dependencies

To be fair, npm squashes these if they're the same version.

**Comment by CrossEye** [comment]
>> What I don't like is importing multiple copies of shared dependencies

> To be fair, npm squashes these if they're the same version.

Perhaps I'm confused by this, as I haven't used it, but would that mean publishing all our internal dependencies as modules too?  We have quite a number of functions meant only for internal use, not published anywhere on their own.

**Comment by krainboltgreene** [comment]
Maybe you don't *have to*, but it'll probably require effort to avoid it. Frankly though I don't think that's a terrible thing. To be fair the two internal functions I've seen would be pretty useful for my libraries. Right now I'm importing the public ramda functions, but really I would love to use one of the internal ones.

**Comment by CrossEye** [comment]
But remember that the internal functions were not written with public consumption in mind.  They don't necessarily have an API that we would want to expose, and their implementations are designed only for the uses we put them to internally.

However, #2161 demonstrates that we may have gone overboard with this.

**Comment by krainboltgreene** [comment]
Packages have versions. It's totally okay to ship internal use plugins because you don't have to care if you change them. That's what the X is for in X.Y.Z.

**Comment by olsonpm** [comment]
Weird - I never even thought of that internal dependency issue, but sure enough lodash publishes those too.  Seems odd, but the advantages are obvious.

**Comment by olsonpm** [comment]
FYI - lodash 5 won't have each utility as a separate npm package.  jdalton is instead minimizing the disk-requirements for his distributable by slimming down the code-base, gzipping the source, then pushing for webpack to both unzip and package only the utilities the application uses.

here's the [lodash v5 roadmap](https://github.com/lodash/lodash/wiki/Roadmap) - I found the direction pretty interesting.

**Comment by krainboltgreene** [comment]
I wonder how much pain is attributed to the fact that it's not an organization on npm.

**Comment by olsonpm** [comment]
Yeah not sure.  Based on his direction [to depend on esm/reify](https://github.com/standard-things/esm/issues/2#issuecomment-304570105) (just found that yesterday), my guess is that he wants to have a single "fits-all" workflow.

**Comment by krainboltgreene** [comment]
That sort of stuff is going into territory that I don't have experience in, but I will say that lerna is not ready to support ramda IMO. They've done an incredible job, but I've had frustrating experiences so far and the team seems very set on a "Lerna Way".

**Comment by kedashoe** [comment]
Don't really have anything to add, just want to say I really appreciate all the work everyone is doing in this area and we will definitely tackle it in the near future.

Do you have any opinions on reify at this point @olsonpm ? I had not heard of it, it looks kinda scary to me.

**Comment by olsonpm** [comment]
I do not.  I honestly just stumbled on it yesterday and don't understand its use-case.  It's trying to get around a build step which is the crux of my confusion.  I just assume webpack is pretty baked into most workflows at this point, hardly with the sole intention of packaging imports into an es5 bundle.

I also disagree with the [author's stance on nested imports](https://github.com/benjamn/reify/blob/master/WHY_NEST_IMPORTS.md), but that ship seems sailed anyway which is fortunate.

As for esm, I'm just interested in what the end-result will look like.  Similar to reify, I don't understand where it might be useful.  It always bothers me how programmers explain "what" so readily, and leave "why" as an exercise for the reader.

**Comment by olsonpm** [comment]
Sorry if this ping is noise @kedashoe - but I'm finally bored of my new job and started back on a personal project again - just found my answer to "why use esm" just now via [jdalton's post (aug 10th)](https://medium.com/web-on-the-edge/es-modules-in-node-today-32cff914e4b) and thought you'd be interested.  Hopefully after a few reads and a night's rest, I'll understand it haha.
